ledger-core/virtual/execute/execute.go

```diff
@@ -329,74 +310,46 @@ func negotiateIsolation(methodIsolation, callIsolation contract.MethodIsolation)
 	return res, nil
 }
 
-func (s *SMExecute) stepDeduplicate(ctx smachine.ExecutionContext) smachine.StateUpdate {
-	duplicate := false
-
-	action := func(state *object.SharedState) {
-		if !state.KnownRequests.Add(s.execution.Isolation.Interference, s.execution.Outgoing) {
-			duplicate = true
-		}
-	}
-
-	switch s.objectSharedState.Prepare(action).TryUse(ctx).GetDecision() {
-	case smachine.NotPassed:
-		return ctx.WaitShared(s.objectSharedState.SharedDataLink).ThenRepeat()
-	case smachine.Impossible:
-		panic(throw.NotImplemented())
-	case smachine.Passed:
-		// go further
-	default:
-		panic(throw.Impossible())
-	}
-
-	if duplicate {
-		ctx.Log().Warn("duplicate found")
-		return ctx.Stop()
-	}
+type DeduplicationAction byte
 
-	if !s.outgoingFromSlotPulse() {
-		return ctx.Jump(s.stepDeduplicateUsingPendingsTable)
-	}
-	return ctx.Jump(s.stepTakeLock)
-}
+const (
+	Stop DeduplicationAction = iota + 1
+	SendResultAndStop
+	DeduplicateThroughPreviousExecutor
+	ContinueExecute
+)
 
-func (s *SMExecute) stepDeduplicateUsingPendingsTable(ctx smachine.ExecutionContext) smachine.StateUpdate {
+func (s *SMExecute) stepDeduplicate(ctx smachine.ExecutionContext) smachine.StateUpdate {
 	var (
-		isDuplicate bool
-		isActive    bool
-		objectState object.State
+		deduplicateAction DeduplicationAction
+		msg               *payload.VCallResult
 	)
-	action := func(state *object.SharedState) {
-		pendingList := state.PendingTable.GetList(s.execution.Isolation.Interference)
-		isActive, isDuplicate = pendingList.GetState(s.execution.Outgoing)
-		objectState = state.GetState()
-	}
 
-	switch s.objectSharedState.Prepare(action).TryUse(ctx).GetDecision() {
-	case smachine.NotPassed:
-		return ctx.WaitShared(s.objectSharedState.SharedDataLink).ThenRepeat()
-	case smachine.Impossible:
-		panic(throw.NotImplemented())
-	case smachine.Passed:
-	default:
-		panic(throw.NotImplemented())
-	}
+	// deduplication algorithm
+	action := func(state *object.SharedState) {
+		deduplicateAction0, msg0 := s.deduplicate(state)
```

Comment:
what?

Reply:
why?

Reply:
deduplicateAction, msg = s.deduplicate(state)

Reply:
Ohh, it is works.

Reply:
Fixed